nil-docs/zkllvm/getting-started/compiling-a-circuit.mdx

```diff
@@ -0,0 +1,193 @@
+import Tabs from '@theme/Tabs'
+import TabItem from '@theme/TabItem'
+
+# Compiling a circuit
+
+This tutorial contains step-by-step instructions on how to compile the [**arithmetic circuit created previously**](./writing-a-simple-circuit).
+
+There are two possible ways for transforming a circuit into a proof binary and Solidity code.
+
+- By calling individual tools
+- By using [**the zkllvm-template project**](https://github.com/NilFoundation/zkllvm-template)
+
+## Calling individual tools
+
+The following diagram shows the flow for converting a `.cpp` or a `.rs` file into a proof binary and Solidity code for deployment.
+
+```mermaid
+%%{
+  init: {
+    'theme': 'base',
+    'themeVariables': {
+      'primaryColor': '#0f0f0f',
+      'primaryTextColor': '#f1f1f1',
+      'primaryBorderColor': '#f1f1f1',
+      'lineColor': '#87B6FC',
+      'secondaryColor': '#f1f1f1',
+      'tertiaryColor': '#2f2f2f',
+    },
+    'flowchart':
+    {
+        'defaultRenderer': 'elk',
+        'curve': 'step'
+    }
+  }
+}%%
+flowchart TB
+    classDef transpilerTitle margin-right: 250px;
+    Transpiler:::transpilerTitle
+
+    CPP(C++ code)
+    RUST(Rust code)
+    CLANG(<code>clang</code>)
+    RUSTCOMP(<code>cargo</code> or <code>rustc</code>)
+    IR(Circuit IR)
+    INPUTS(Circuit inputs)
+    ASSIGNER(Assigner)
+    ASSIGNERRESULT(<code>.tbl</code> and <code>.crct</code> files)
+    TRANSPILER(<code>transpiler</code>)
+    PROOFBIN(Proof binary)
+    SOLIDITY(Solidity contracts)
+    CPP --> CLANG
+    RUST --> RUSTCOMP
+    CLANG --> IR
+    RUSTCOMP --> IR
+    IR --> ASSIGNER
+    INPUTS --> ASSIGNER
+    ASSIGNER --> ASSIGNERRESULT
+    ASSIGNERRESULT --> TRANSPILER
+    TRANSPILER --> PROOFBIN
+    TRANSPILER --> SOLIDITY
+    subgraph Code
+    CPP
+    RUST
+    end
+    subgraph Compilation
+    direction LR
+    CLANG
+    RUSTCOMP
+    IR
+    end
+    subgraph Assigner
+    direction LR
+    INPUTS
+    ASSIGNER
+    ASSIGNERRESULT
+    end
+    subgraph Transpiler
+    direction LR
+    TRANSPILER
+    PROOFBIN
+    SOLIDITY
+    end
+```
+
+Complete the below steps to follow this flow.
+
+### Running the compiler
+
+To generate the circuit IR, do the following.
+
+<Tabs groupId='language'>
```

Comment:
For me this way of structuring is questionable:

1. This tables look awesome when we are giving some small examples in two languages at the same time. But here Rust and C++ have pretty different usages and these "tabs" just looks different. Maybe we can reorganize it? I suggest the following: we can state at this point, that first step is compilation, which translates your source code into LLVM assembly. Next we provide two links for "Compiling C++ code" and "Compiling Rust code". As a result, both pages describe a way to get an `.ll` file. Next, in assigner step we say "Take the file that you generated on previous step". For me it looks a bit cleaner because at this point I guess no one will be interested in both pipelines at the same time: user either code in C++ or in Rust.
2. Description of Rust pipeline is a bit rough. I would suggest we emphasize that building for circuits is almost the same as your normal build process. With that in mind I would try to avoid encouraging users to use `rustc` directly. Once again, you almost never use it without Cargo and I don't see any reasons why you have to with this type of projects. So I'd stick with Cargo and go on with something like "Ok, let's say you have your typical Rust project. All you have to do is just build it for the right target: `assigner-unknown-unknown`. The rest of your Rust-dev experience remains the same." We also probably have to mention here, where user can find generated `.ll` file  — and again it is the same as when compiling for native platform. The same location of output file, just the different extension.
3. We can mention bare `rustc` way just in case somebody gets curios, but it shouldn't be considered as a "normal" option. It's just for some exotic cases. And I think, that if user is capable of using bare Rust compiler, then he would be able to figure out all the arguments he have to provide to it. After all, this is really not the theme for "Getting started".

Reply:
sure, I've changed the structure and the pipeline description. however, I left the explanation of the arguments as is as some people switching to Rust from C++ (as was the case for me) could call `rustc` directly purely out of old habits

Reply:
Okay makes sense